aiidalab/app.py

```diff
@@ -641,29 +645,30 @@ def check_path_exists_changed():
                     self._observer.stop()
 
             self._monitor_thread = Thread(target=check_path_exists_changed)
-            self._monitor_thread.stop_flag = False  # type: ignore
+            self._monitor_thread_stop.clear()
             self._monitor_thread.start()
 
         self._started = True
 
     def stop(self):
         """Stop watching the app repository for file system events."""
         if self._monitor_thread is not None:
-            self._monitor_thread.stop_flag = True
+            self._monitor_thread_stop.set()
 
     def is_alive(self):
         """Return True if this watch is still alive."""
         return self._monitor_thread and self._monitor_thread.is_alive()
 
     def join(self, timeout=None):
-        """Join the watch after stopping.
+        """Join the watch and observer after stopping.
 
         This function will timeout if a timeout argument is provided. Use the
         is_alive() function to determien whether the watch was stopped within
         the given timout.
         """
-        if self._monitor_thread is not None:
+        if self._monitor_thread is not None and self._observer is not None:
             self._monitor_thread.join(timeout=timeout)
+            self._observer.join(timeout=timeout)
```

Comment:
I think it is safer to split this condition, otherwise you risk that the monitor thread started but observer thread is not started, and the join statements will not execute.

```suggestion
        if self._monitor_thread is not None:
            self._monitor_thread.join(timeout=timeout)
        if self._observer is not None:
            self._observer.join(timeout=timeout)
```